hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/clustering/ClusteringPlanOperator.java

```diff
@@ -0,0 +1,139 @@
+
+  public ClusteringPlanOperator(Configuration conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void open() throws Exception {
+    super.open();
+    this.table = FlinkTables.createTable(conf, getRuntimeContext());
+    // when starting up, rolls back all the inflight clustering instants if there exists,
+    // these instants are in priority for scheduling task because the clustering instants are
+    // scheduled from earliest(FIFO sequence).
+    ClusteringUtil.rollbackClustering(table, StreamerUtil.createWriteClient(conf, getRuntimeContext()));
+  }
+
+  @Override
+  public void processElement(StreamRecord<Object> streamRecord) {
+    // no operation
+  }
+
+  @Override
+  public void notifyCheckpointComplete(long checkpointId) {
+    try {
+      table.getMetaClient().reloadActiveTimeline();
+      scheduleClustering(table, checkpointId);
+    } catch (Throwable throwable) {
+      // make it fail-safe
+      LOG.error("Error while scheduling clustering plan for checkpoint: " + checkpointId, throwable);
+    }
+  }
+
+  private void scheduleClustering(HoodieFlinkTable<?> table, long checkpointId) {
+    // the first instant takes the highest priority.
+    Option<HoodieInstant> firstRequested = Option.fromJavaOptional(
+        ClusteringUtils.getPendingClusteringInstantTimes(table.getMetaClient()).stream()
+            .filter(instant -> instant.getState() == HoodieInstant.State.REQUESTED).findFirst());
+    if (!firstRequested.isPresent()) {
+      // do nothing.
+      LOG.info("No clustering plan for checkpoint " + checkpointId);
+      return;
+    }
+
+    String clusteringInstantTime = firstRequested.get().getTimestamp();
+
+    // generate clustering plan
+    // should support configurable commit metadata
+    Option<Pair<HoodieInstant, HoodieClusteringPlan>> clusteringPlanOption = ClusteringUtils.getClusteringPlan(
+        table.getMetaClient(), HoodieTimeline.getReplaceCommitRequestedInstant(clusteringInstantTime));
+
+    if (!clusteringPlanOption.isPresent()) {
+      // do nothing.
+      LOG.info("No clustering plan scheduled");
+      return;
+    }
+
+    HoodieClusteringPlan clusteringPlan = clusteringPlanOption.get().getRight();
+
+    if (clusteringPlan == null || (clusteringPlan.getInputGroups() == null)
+        || (clusteringPlan.getInputGroups().isEmpty())) {
+      // do nothing.
+      LOG.info("Empty clustering plan for instant " + clusteringInstantTime);
+    } else {
+      HoodieInstant instant = HoodieTimeline.getReplaceCommitRequestedInstant(clusteringInstantTime);
+      // Mark instant as clustering inflight
+      table.getActiveTimeline().transitionReplaceRequestedToInflight(instant, Option.empty());
+      table.getMetaClient().reloadActiveTimeline();
+
+      for (HoodieClusteringGroup clusteringGroup : clusteringPlan.getInputGroups()) {
+        LOG.info("ClusteringPlanSourceFunction cluster " + clusteringGroup + " files");
+        output.collect(new StreamRecord<>(
```

Comment:
The message `ClusteringPlanSourceFunction` is wrong, `Execute clustering plan for instant {} as {} file slices` ?